File: sandbox/plc4c/spi/src/write.c
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <plc4c/plc4c.h>
#include <plc4c/spi/types_private.h>
#include <plc4c/write.h>
#include <stdlib.h>

plc4c_connection *plc4c_write_request_get_connection(
    plc4c_write_request *write_request) {
  return write_request->connection;
}

void plc4c_write_request_set_connection(plc4c_write_request *write_request,
                                        plc4c_connection *connection) {
  write_request->connection = connection;
}

plc4c_return_code plc4c_write_request_add_item(
    plc4c_write_request *write_request, char *address, plc4c_data *value) {
  // Parse an address string and get a driver-dependent data-structure
  // representing the address back.
  plc4c_request_value_item *value_item;
  plc4c_item *address_item;
  plc4c_return_code result;

  address_item = malloc(sizeof(plc4c_item));
  if(address_item == NULL) 
    return NO_MEMORY;
  
  result = write_request->connection->driver->parse_address_function(
          address, &address_item->address);
  
  if(result != OK) 
    return result;
  
  // Create a new value item, binding an address item to a value.
  value_item = malloc(sizeof(plc4c_request_value_item));
  value_item->item = address_item;
  value_item->value = value;

  plc4c_utils_list_insert_head_value(write_request->items, value_item);
  return OK;
}

plc4c_return_code plc4c_write_request_execute(
    plc4c_write_request *write_request,
    plc4c_write_request_execution **write_request_execution) {

  plc4c_write_request_execution *new_write_request_execution;
  plc4c_connection *connection;
  plc4c_driver *driver;
  plc4c_system *system;
  plc4c_list *system_tasks;

  // Extract some plc4c pointers for clarity
  connection = plc4c_write_request_get_connection(write_request);
  system = plc4c_connection_get_system(connection);
  driver = plc4c_connection_get_driver(connection);
  system_tasks = plc4c_system_get_task_list(system);

  // Inject the default write context into the system task.
  new_write_request_execution = malloc(sizeof(plc4c_write_request_execution));
  new_write_request_execution->write_request = write_request;
  new_write_request_execution->write_response = NULL;
  new_write_request_execution->system_task = NULL;
  
  driver->write_function(new_write_request_execution, &(new_write_request_execution->system_task));

  // Increment the number of running tasks for this connection.
  plc4c_connection_task_added(connection);

  // Add the new task to the task-list.
  plc4c_utils_list_insert_tail_value(system_tasks, new_write_request_execution->system_task);

  *write_request_execution = new_write_request_execution;
  return OK;
}

bool plc4c_write_request_check_finished_successfully(
    plc4c_write_request_execution *write_request_execution) {

  if (write_request_execution == NULL) {
    return true;
  }
  if (write_request_execution->system_task == NULL) {
    return true;
  }
  return write_request_execution->system_task->completed;
}

bool plc4c_write_request_execution_check_completed_with_error(
    plc4c_write_request_execution *write_request_execution) {
  return false;
}

plc4c_write_response *plc4c_write_request_execution_get_response(
    plc4c_write_request_execution *write_request_execution) {
  if (write_request_execution == NULL) {
    return NULL;
  }
  return write_request_execution->write_response;
}




void plc4c_write_request_destroy(plc4c_write_request *write_request) {
  write_request->connection->driver->free_write_request_function(write_request);
  free(write_request);
}

void plc4c_write_request_execution_destroy(
    plc4c_write_request_execution *write_request_execution) {
  free(write_request_execution);
}

void plc4c_write_response_destroy(
    plc4c_write_response *write_response) {
  write_response->write_request->connection->driver
      ->free_write_response_function(write_response);
  free(write_response);
}

File: plc4go/examples/ads/write/Write.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package main

import (
	"time"

	plc4go "github.com/apache/plc4x/plc4go/pkg/api"
	"github.com/apache/plc4x/plc4go/pkg/api/drivers"
	apiValues "github.com/apache/plc4x/plc4go/pkg/api/values"
	spiValues "github.com/apache/plc4x/plc4go/spi/values"

	"github.com/rs/zerolog/log"
)

func main() {
	driverManager := plc4go.NewPlcDriverManager()
	drivers.RegisterAdsDriver(driverManager)
	connectionChan := driverManager.GetConnection("ads:tcp://192.168.23.20?sourceAmsNetId=192.168.23.200.1.1&sourceAmsPort=65534&targetAmsNetId=192.168.23.20.1.1&targetAmsPort=851")
	connection := <-connectionChan

	duration, _ := time.ParseDuration("1.234S")
	lduration, _ := time.ParseDuration("24015H23M12.034002044S")
	date, _ := time.Parse("2014-11-12", "1978-03-28")
	timeOfDay, _ := time.Parse("11:45:26.371Z", "15:36:30.123")
	dateAndTime, _ := time.Parse("2014-11-12T11:45:26.371Z", "1996-05-06T15:36:30")

	// Prepare the data structure for a custom type.
	children := map[string]apiValues.PlcValue{}
	children["hurz_BOOL"] = spiValues.NewPlcBOOL(true)
	children["hurz_BYTE"] = spiValues.NewPlcBYTE(1)
	children["hurz_WORD"] = spiValues.NewPlcWORD(2)
	children["hurz_DWORD"] = spiValues.NewPlcDWORD(3)
	children["hurz_LWORD"] = spiValues.NewPlcLWORD(4)
	children["hurz_SINT"] = spiValues.NewPlcSINT(5)
	children["hurz_USINT"] = spiValues.NewPlcUSINT(6)
	children["hurz_INT"] = spiValues.NewPlcINT(7)
	children["hurz_UINT"] = spiValues.NewPlcUINT(8)
	children["hurz_DINT"] = spiValues.NewPlcDINT(9)
	children["hurz_UDINT"] = spiValues.NewPlcUDINT(10)
	children["hurz_LINT"] = spiValues.NewPlcLINT(11)
	children["hurz_ULINT"] = spiValues.NewPlcULINT(12)
	children["hurz_REAL"] = spiValues.NewPlcREAL(13.0)
	children["hurz_LREAL"] = spiValues.NewPlcLREAL(14.0)
	children["hurz_STRING"] = spiValues.NewPlcSTRING("hurz")
	children["hurz_WSTRING"] = spiValues.NewPlcWSTRING("wolf")
	children["hurz_TIME"] = spiValues.NewPlcTIME(duration)
	children["hurz_LTIME"] = spiValues.NewPlcLTIME(lduration)
	children["hurz_DATE"] = spiValues.NewPlcDATE(date)
	children["hurz_TIME_OF_DAY"] = spiValues.NewPlcTIME_OF_DAY(timeOfDay)
	children["hurz_DATE_AND_TIME"] = spiValues.NewPlcDATE_AND_TIME(dateAndTime)
	writeRequest, err := connection.GetConnection().WriteRequestBuilder().
		AddTagAddress("value-bool", "MAIN.hurz_BOOL", spiValues.NewPlcBOOL(true)).                   // 1
		AddTagAddress("value-byte", "MAIN.hurz_BYTE", spiValues.NewPlcBYTE(42)).                     // 1
		AddTagAddress("value-word", "MAIN.hurz_WORD", spiValues.NewPlcWORD(42424)).                  // 2
		AddTagAddress("value-dword", "MAIN.hurz_DWORD", spiValues.NewPlcDWORD(4242442424)).          // 4
		AddTagAddress("value-lword", "MAIN.hurz_LWORD", spiValues.NewPlcLWORD(4242442424242424242)). // 8
		AddTagAddress("value-sint", "MAIN.hurz_SINT", spiValues.NewPlcSINT(-42)).                    // 1
		AddTagAddress("value-usint", "MAIN.hurz_USINT", spiValues.NewPlcUSINT(42)).                  // 1
		AddTagAddress("value-int", "MAIN.hurz_INT", spiValues.NewPlcINT(-2424)).                     // 2
		AddTagAddress("value-uint", "MAIN.hurz_UINT", spiValues.NewPlcUINT(42424)).                  // 2
		AddTagAddress("value-dint", "MAIN.hurz_DINT", spiValues.NewPlcDINT(-242442424)).             // 4
		AddTagAddress("value-udint", "MAIN.hurz_UDINT", spiValues.NewPlcUDINT(4242442424)).          // 4
		AddTagAddress("value-lint", "MAIN.hurz_LINT", spiValues.NewPlcLINT(-4242442424242424242)).   // 8
		AddTagAddress("value-ulint", "MAIN.hurz_ULINT", spiValues.NewPlcULINT(4242442424242424242)). // 8
		AddTagAddress("value-real", "MAIN.hurz_REAL", spiValues.NewPlcREAL(3.14159265359)).          // 4
		AddTagAddress("value-lreal", "MAIN.hurz_LREAL", spiValues.NewPlcLREAL(2.71828182846)).       // 8
		AddTagAddress("value-string", "MAIN.hurz_STRING", spiValues.NewPlcSTRING("hurz")).           // 4
		AddTagAddress("value-wstring", "MAIN.hurz_WSTRING", spiValues.NewPlcWSTRING("wolf")).        // 8
		AddTagAddress("value-time", "MAIN.hurz_TIME", spiValues.NewPlcTIME(duration)).
		AddTagAddress("value-ltime", "MAIN.hurz_LTIME", spiValues.NewPlcLTIME(lduration)).
		AddTagAddress("value.date", "MAIN.hurz_DATE", spiValues.NewPlcDATE(date)).
		AddTagAddress("value-time-of-day", "MAIN.hurz_TIME_OF_DAY", spiValues.NewPlcTIME_OF_DAY(timeOfDay)).
		AddTagAddress("value-date-and-time", "MAIN.hurz_DATE_AND_TIME", spiValues.NewPlcDATE_AND_TIME(dateAndTime)).
		AddTagAddress("value-struct", "MAIN.hurz_Struct", spiValues.NewPlcStruct(children)).
		Build()
	if err != nil {
		panic(err)
	}
	writeResponseChannel := writeRequest.Execute()
	writeResult := <-writeResponseChannel
	if writeResult.GetErr() != nil {
		log.Error().Err(writeResult.GetErr()).Msg("error in response")
		return
	}
	writeResponse := writeResult.GetResponse()
	for _, tagName := range writeResponse.GetTagNames() {
		responseCode := writeResponse.GetResponseCode(tagName)
		print(responseCode)
	}
}
